server/src/validation.ts

```diff
@@ -316,3 +318,65 @@ export function validateAvailabilityInput(data: any): AvailabilityInput {
 
   return data;
 }
+
+let stateLookup: Map<Lowercase<string>, typeof states[number]>;
+
+/**
+ * Get the correct state abbreviation given the full name or some type of
+ * well-known abbreviation for the state. If no matching state can be found,
+ * this will throw `ValueError`.
+ */
+export function validateState(input: string): string {
+  if (!stateLookup) {
+    stateLookup = new Map();
+    const references: Array<keyof typeof states[number]> = [
+      "name",
+      "iso",
+      "usps",
+      "gpo",
+      "ap",
+    ];
+    for (const state of states) {
+      for (const key of references) {
+        const reference = state[key] as string;
+        if (reference) {
+          stateLookup.set(reference.toLowerCase(), state);
+        }
+      }
+    }
+  }
+
+  const state = stateLookup.get(input.trim().toLowerCase());
+  if (state) {
+    return state.usps;
+  }
+  throw new ValueError(`Unknown state: "${input}"`);
+}
+
+export function validateLocationInput(
+  data: any,
+  requiredFields = false
+): ProviderLocationInput {
+  const result = { ...data };
+  delete result.availability;
+  delete result.created_at;
+  delete result.updated_at;
+
+  if (result.state) {
+    result.state = validateState(data.state);
+  }
+
+  // TODO: Validate rest of schema
+  if (requiredFields) {
+    const missing = ["name", "provider", "state"].filter(
+      (field) => !result[field]
+    );
+    if (missing.length) {
+      throw new ValueError(
+        `The location is missing values for: ${missing.join(", ")}`
+      );
+    }
+  }
```

Comment:
I got a bit lazy here, I guess. Not sure if I should go ahead and add a schema object for this and validate that. (Somewhat complex because we need two — one that requires some fields and one that doesn’t, since updates don’t have any real required fields.)

Reply:
Sure, not clear you need to go through a lot of effort here to validate things further.

But if you did want to, just curious, any easy way to construct a derivative or child validator such that you could make the looser one and then overlay the required fields on top for the second?

Reply:
Went ahead and did this. Hopefully won’t cause any issues on real data! 🤞